### PR TITLE
Subscription management

### PR DESCRIPTION
**Pull request by willmero**

### CODE REVIEW

File: clientSuccessClient.js [hunk 1]
```diff
@@ -407,14 +409,103 @@ JS.class(ClientSuccessClient, {
...
 		},
+
+		/**
+		 * Get ClientSuccess product ID based on product name
+		 * @param  {String} productName - Name of the Product
+		 * @return {Integer}            - ID of the Product
+		 */
+		getProductID : async function(productName) {
+			const clientSuccessProducts = await this.hitClientSuccessAPI('GET', 'products');
+			for (let x = 0; x < clientSuccessProducts.length; x++) {
```
Comment (juuxstar) [inline clientSuccessClient.js]: Generally, use built-in looping method (either lodash or native).  This looks to me like a great candidate for `find`

File: clientSuccessClient.js [hunk 2]
```diff
@@ -407,14 +409,103 @@ JS.class(ClientSuccessClient, {
...
+			throw new CustomError({ status : 404, message : 'Product not found' });
+		},
+
+		/**
+		 * Create a new ClientSuccess Product Type
+		 * @param  {String}  name               - Name of the new Product Type
+		 * @param  {Boolean} [recurring = true] - Denotes if Subscription is recurring or not
+		 * @return Promise<Object>              - Resulting ClientSuccess Subscription object
+		 */
+		createProductType : async function({ name = undefined, recurring = true } = {}) {
```
Comment (juuxstar) [inline clientSuccessClient.js]: Why is the `name` parameter an option with a default value of undefined when clearly the function requires it.

I'd change the signature to: `async function(name, { recurring=true })`

File: clientSuccessClient.js [hunk 3]
```diff
@@ -407,14 +409,103 @@ JS.class(ClientSuccessClient, {
...
+					return clientSuccessProducts[x].id;
+				}
+			}
+			throw new CustomError({ status : 404, message : 'Product not found' });
+		},
+
+		/**
+		 * Create a new ClientSuccess Product Type
+		 * @param  {String}  name               - Name of the new Product Type
+		 * @param  {Boolean} [recurring = true] - Denotes if Subscription is recurring or not
```
Comment (juuxstar) [inline clientSuccessClient.js]: See our code conventions for how to document the option parameter.

https://github.com/Roadmunk/roadmunk/wiki/Coding-Conventions#optional-function-parameters

Should be:
```
@param {Object} [options] ...
@param {Boolean} [options.recurring=true] ...
```

File: clientSuccessClient.js [hunk 4]
```diff
@@ -407,14 +409,103 @@ JS.class(ClientSuccessClient, {
...
+		/**
+		 * Get all subscription items for a client
+		 * @param  {String} clientID - ClientSuccess Client ID
+		 * @return Promise<Object>   - Subscriptions list for the provided Client ID
+		 */
+		getClientActiveSubscriptions : async function(clientID) {
+			this.validateClientSuccessId(clientID);
+
+			const clientSubscriptions = await this.hitClientSuccessAPI('GET', `subscriptions?clientId=${clientID}`);
+			if (clientSubscriptions.length > 0) {
```
Comment (juuxstar) [inline clientSuccessClient.js]: Reverse your if condition and bail early with the CustomError.

Keeps the "main" logic of the function within the main body with a natural return at the end.  Also keeps the majority of the code from being indented one more time.

File: clientSuccessClient.js [hunk 5]
```diff
@@ -407,14 +409,103 @@ JS.class(ClientSuccessClient, {
...
+				// ClientSuccess has outlined that Subscriptions that are considered 'Active' have attribute isPotential = false
+				// Loop through the returned array and pull out the 'Active' subscriptions
+				const activeSubscriptions = [];
+				for (let i = 0; i < clientSubscriptions.length; i++) {
```
Comment (juuxstar) [inline clientSuccessClient.js]: use a `forEach` here.  It's easier to read and less prone to off-by-one errors.

Comment (juuxstar) [reply clientSuccessClient.js]: Or better yet, use a `_.reduce` here.

File: clientSuccessClient.js [hunk 6]
```diff
@@ -407,14 +409,103 @@ JS.class(ClientSuccessClient, {
...
+				throw new CustomError({ status : 400, message : 'Product ID Required for Deletion' });
+			}
+
+			return this.hitClientSuccessAPI('DELETE', `products/${productId}`);
+		},
+
+		/**
+		 * Get all subscription items for a client
+		 * @param  {String} clientID - ClientSuccess Client ID
+		 * @return Promise<Object>   - Subscriptions list for the provided Client ID
```
Comment (juuxstar) [inline clientSuccessClient.js]: -> `Promise<Object[]>`

File: clientSuccessClient.js [hunk 7]
```diff
@@ -407,14 +409,103 @@ JS.class(ClientSuccessClient, {
...
+		 * @param  {Object} attributes - List of Subscription attributes
+		 * @return Promise<Object>     - Resulting ClientSuccess Subscription created
+		 */
+		createClientSubscription : async function(clientID, attributes) {
+			this.validateClientSuccessId(clientID);
+
+			// add clientID to the attributes array
+			const finalAttributes = Object.assign(attributes, { clientId : clientID });
+
+			const clientSubscription = await this.hitClientSuccessAPI('POST', 'subscriptions', finalAttributes);
```
Comment (juuxstar) [inline clientSuccessClient.js]: just `return this.hitClientSuccessAPI()`...

File: clientSuccessClient.js [hunk 8]
```diff
@@ -407,14 +409,103 @@ JS.class(ClientSuccessClient, {
...
 	},
 });
 
 JS.class(CustomError, {
 	inherits : Error,
 
-	constructor : function({ status, message }) {
-		this.status  = status;
-		this.message = message;
+	constructor : function({ status, message, userMessage }) {
```
Comment (juuxstar) [inline clientSuccessClient.js]: Declare your class fields.  Helps with documentation.

File: test/tests.js [hunk 9]
```diff
@@ -738,6 +738,99 @@ describe('clientSuccessClient', function() {
...
+		});
+	});
+
+	describe('createClientSubscription', async function() {
+		this.timeout(15000);
+		let testClient;
+
+		before(async function() {
+			// create a test client with initial attributes to use in updateClient tests
+			const testClientName = `${(new Date()).getTime()}`;
```
Comment (juuxstar) [inline test/tests.js]: Why have the separate const declaration?  Can simplify by inlining the expression into where it's being used.

**Comment by juuxstar** [comment]
Please also add our show eslint.  Simple instructions at https://github.com/Roadmunk/eslint-config-roadmunk.

File: clientSuccessClient.js (hunk 5 above)
Comment (willmero) [reply clientSuccessClient.js]: @juuxstar went with forEach, was having difficulties getting `_.reduce` to work properly. From the docs it looks like `_.reduce` is used to take an array with mixed values and matching keys, and summing them down into a single value. Here, we are looking to filter out objects.

File: clientSuccessClient.js (hunk 6 above)
Comment (willmero) [reply clientSuccessClient.js]: Yes oops! Multiple subs!

File: clientSuccessClient.js [hunk 10]
```diff
@@ -407,14 +409,132 @@ JS.class(ClientSuccessClient, {
...
+		 * Create a ClientSuccess Subscription line item under the passed Client
+		 * @param  {String} clientID   - ClientSuccess Client
+		 * @param  {Object} attributes - List of Subscription attributes
+		 * @return Promise<Object>     - Resulting ClientSuccess Subscription created
+		 */
+		createClientSubscription : async function(clientID, attributes) {
+			this.validateClientSuccessId(clientID);
+
+			// add clientID to the attributes array
+			const finalAttributes = Object.assign(attributes, { clientId : clientID });
```
[the remark this answers is not in the document]
Comment (willmero) [reply clientSuccessClient.js]: @juuxstar this is a great point... Didn't think about that!

File: test/tests.js [hunk 11]
```diff
@@ -738,6 +742,111 @@ describe('clientSuccessClient', function() {
...
+			testClient = await CS.createClient(testClientAttributesInitial);
+		});
+
+		after(async function() {
+			CS.closeClient(testClient.id);
+		});
+
+		it('should successfully create a ClientSuccess subscription', async function() {
+			const testProductId = await CS.getProductId('Collaborators');
+			// console.log(`Creating subscription under product ID: ${testProductId} with client ID: ${testClient.id}`);
```
[the remark this answers is not in the document]
Comment (willmero) [reply test/tests.js]: Meh, I was going to leave it in for future debugging... but, I'll just not be lazy and re-write the line if I need it 🙊

**Pull request merged**